File: scannerService.js
const EventEmitter = require('events');
const { SerialPort } = require('serialport');

const SCAN_IDLE_FLUSH_MS = 80;
const DEBUG_CHUNK_LIMIT = 20;

function normalizeScanText(rawValue) {
  return String(rawValue || '')
    .replace(/[\u0000-\u001f\u007f]/g, '')
    .trim();
}

function toHexString(buffer) {
  return Array.from(buffer || [])
    .map((value) => value.toString(16).padStart(2, '0').toUpperCase())
    .join(' ');
}

function toTextPreview(buffer) {
  return String((buffer || Buffer.alloc(0)).toString('utf8') || '')
    .replace(/\r/g, '<CR>')
    .replace(/\n/g, '<LF>');
}

class ScannerError extends Error {
  constructor(message, cause) {
    super(message);
    this.name = 'ScannerError';
    this.cause = cause;
  }
}

class ScannerService extends EventEmitter {
  constructor() {
    super();
    this.port = null;
    this.currentConfig = null;
    this.latestScan = null;
    this.consumedScanId = null;
    this.buffer = '';
    this.flushTimer = null;
    this.debugState = this.createDebugState();
  }

  createDebugState() {
    return {
      bytesReceived: 0,
      chunksReceived: 0,
      lastChunkAt: null,
      lastPublishedAt: null,
      modemSignals: null,
      bufferPreview: '',
      recentChunks: []
    };
  }

  async configure(config, onScan) {
    this.currentConfig = {
      comPort: config.comPort,
      baudrate: Number(config.baudrate),
      dataBits: Number(config.dataBits),
      parity: String(config.parity || 'none').toLowerCase(),
      stopBits: Number(config.stopBits),
      timeoutMs: Number(config.timeoutMs || 5000),
      dtr: config.dtr !== false,
      rts: config.rts !== false,
      enabled: Boolean(config.enabled)
    };

    if (typeof onScan === 'function') {
      this.removeAllListeners('scan');
      this.on('scan', onScan);
    }

    if (!this.currentConfig.enabled) {
      await this.disconnect();
      return { connected: false, enabled: false };
    }

    this.debugState = this.createDebugState();
    return this.connect();
  }

  async connect() {
    try {
      await this.disconnect();

      this.port = new SerialPort({
        path: this.currentConfig.comPort,
        baudRate: this.currentConfig.baudrate,
        dataBits: this.currentConfig.dataBits,
        stopBits: this.currentConfig.stopBits,
        parity: this.currentConfig.parity,
        autoOpen: false
      });

      await new Promise((resolve, reject) => {
        this.port.open((error) => {
          if (error) {
            reject(error);
            return;
          }

          resolve();
        });
      });

      await this.applyLineSignals();

      this.port.on('data', (chunk) => {
        this.handleIncomingChunk(chunk);
      });

      this.port.on('error', (error) => {
        console.error('[scanner] serial error', error);
      });

      console.log(`[scanner] connected ${this.currentConfig.comPort}`);
      return { connected: true, enabled: true };
    } catch (error) {
      console.error('[scanner] connect failed', error);
      throw new ScannerError('Scanner serial connect failed', error);
    }
  }

  async disconnect() {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }

    if (!this.port) {
      return;
    }

    const port = this.port;
    this.port = null;
    this.buffer = '';
    this.debugState.bufferPreview = '';

    if (!port.isOpen) {
      return;
    }

    await new Promise((resolve) => {
      port.close(() => resolve());
    });
  }

  getLatestScan() {
    return this.latestScan;
  }

  syncLatestScan(scanEvent) {
    if (!scanEvent || !scanEvent.rawText) {
      return;
    }

    if (scanEvent.id && this.consumedScanId && scanEvent.id !== this.consumedScanId) {
      this.consumedScanId = null;
    }

    this.latestScan = {
      id: scanEvent.id || null,
      rawText: scanEvent.rawText,
      scannedAt: scanEvent.scannedAt || new Date().toISOString()
    };
    this.debugState.lastPublishedAt = this.latestScan.scannedAt;
  }

  isConsumedScan(scanEvent) {
    return Boolean(
      scanEvent &&
      scanEvent.id &&
      this.consumedScanId &&
      scanEvent.id === this.consumedScanId
    );
  }

  getLatestVisibleScan(fallbackScan = null) {
    if (this.latestScan) {
      return this.latestScan;
    }

    if (this.isConsumedScan(fallbackScan)) {
      return null;
    }

    return fallbackScan || null;
  }

  clearLatestScan(match = null) {
    if (!this.latestScan) {
      return;
    }

    if (!match) {
      this.latestScan = null;
      return;
    }

    const idMatched = match.scannerEventId && this.latestScan.id === match.scannerEventId;
    const qrMatched = !this.latestScan.id && match.qrCode && this.latestScan.rawText === match.qrCode;
    if (idMatched || qrMatched) {
      this.latestScan = null;
    }
  }

  markScanConsumed(match = null) {
    if (match && match.scannerEventId) {
      this.consumedScanId = match.scannerEventId;
    }

    this.clearLatestScan(match);
  }

  consumeCurrentScan(match = null) {
    if (match && match.scannerEventId) {
      this.consumedScanId = match.scannerEventId;
    } else if (this.latestScan && this.latestScan.id) {
      this.consumedScanId = this.latestScan.id;
    }

    this.latestScan = null;
    this.buffer = '';
    this.debugState.bufferPreview = '';
  }

  getDebugState() {
    return {
      connected: this.isConnected(),
      latestScan: this.latestScan,
      currentConfig: this.currentConfig,
      debug: {
        ...this.debugState,
        recentChunks: this.debugState.recentChunks.slice()
      }
    };
  }

  async updateLineSignals(options) {
    if (!this.currentConfig) {
      throw new ScannerError('Scanner config not initialized');
    }

    this.currentConfig.dtr = Boolean(options.dtr);
    this.currentConfig.rts = Boolean(options.rts);

    if (options.reconnect !== false) {
      await this.connect();
    } else if (this.port && this.port.isOpen) {
      await this.applyLineSignals();
    }

    return this.getDebugState();
  }

  isConnected() {
    return Boolean(this.port && this.port.isOpen);
  }

  async applyLineSignals() {
    await new Promise((resolve, reject) => {
      this.port.set({
        dtr: this.currentConfig.dtr !== false,
        rts: this.currentConfig.rts !== false,
        brk: false
      }, (error) => {
        if (error) {
          reject(error);
          return;
        }

        resolve();
      });
    });

    await new Promise((resolve) => {
      this.port.get((error, status) => {
        if (error) {
          console.error('[scanner] get modem status failed', error);
          resolve();
          return;
        }

        this.debugState.modemSignals = status;
        resolve();
      });
    });
  }

  handleIncomingChunk(chunk) {
    const bufferChunk = Buffer.isBuffer(chunk)
      ? chunk
      : Buffer.from(String(chunk || ''), 'utf8');
    const textChunk = Buffer.isBuffer(chunk)
      ? chunk.toString('utf8')
      : String(chunk || '');

    this.captureDebugChunk(bufferChunk);
    this.buffer += textChunk;
    this.debugState.bufferPreview = this.buffer
      .replace(/\r/g, '<CR>')
      .replace(/\n/g, '<LF>')
      .slice(-200);

    let boundaryIndex = this.findBoundaryIndex(this.buffer);
    while (boundaryIndex !== -1) {
      const payload = this.buffer.slice(0, boundaryIndex);
      this.buffer = this.buffer.slice(boundaryIndex + 1);
      this.publishScan(payload);
      boundaryIndex = this.findBoundaryIndex(this.buffer);
    }

    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
    }

    this.flushTimer = setTimeout(() => {
      const payload = this.buffer;
      this.buffer = '';
      this.debugState.bufferPreview = '';
      this.publishScan(payload);
    }, SCAN_IDLE_FLUSH_MS);
  }

  captureDebugChunk(bufferChunk) {
    this.debugState.bytesReceived += bufferChunk.length;
    this.debugState.chunksReceived += 1;
    this.debugState.lastChunkAt = new Date().toISOString();
    this.debugState.recentChunks.push({
      at: this.debugState.lastChunkAt,
      size: bufferChunk.length,
      hex: toHexString(bufferChunk),
      textPreview: toTextPreview(bufferChunk)
    });

    if (this.debugState.recentChunks.length > DEBUG_CHUNK_LIMIT) {
      this.debugState.recentChunks = this.debugState.recentChunks.slice(-DEBUG_CHUNK_LIMIT);
    }
  }

  findBoundaryIndex(bufferText) {
    if (!bufferText) {
      return -1;
    }

    const carriageReturn = bufferText.indexOf('\r');
    const lineFeed = bufferText.indexOf('\n');

    if (carriageReturn === -1) {
      return lineFeed;
    }

    if (lineFeed === -1) {
      return carriageReturn;
    }

    return Math.min(carriageReturn, lineFeed);
  }

  publishScan(rawValue) {
    const rawText = normalizeScanText(rawValue);
    if (!rawText) {
      return;
    }

    this.latestScan = {
      rawText,
      scannedAt: new Date().toISOString()
    };
    this.debugState.lastPublishedAt = this.latestScan.scannedAt;

    console.log(`[scanner] scan received ${rawText}`);
    this.emit('scan', this.latestScan);
  }
}

module.exports = {
  ScannerError,
  scannerService: new ScannerService()
};

File: src/ATEQ.LeakTest.Web/Services/ScannerService.cs
using System.Collections.Concurrent;
using System.IO.Ports;
using ATEQ.LeakTest.Web.Models;

namespace ATEQ.LeakTest.Web.Services;

public class ScannerService
{
    private const int ScanIdleFlushMs = 80;
    private const int DebugChunkLimit = 20;
    private readonly FeatureFlags _flags;

    public ScannerService(FeatureFlags flags) { _flags = flags; }

    private SerialPort? _port;
    private CommConfig? _currentConfig;
    private CancellationTokenSource? _readCts;
    private readonly object _lock = new();

    // Scan state
    private ScannerEvent? _latestScan;
    private string? _consumedScanId;
    private string _buffer = "";

    // Debug state
    private long _bytesReceived;
    private long _chunksReceived;
    private string? _lastChunkAt;
    private string? _lastPublishedAt;
    private object? _modemSignals;
    private string _bufferPreview = "";
    private readonly ConcurrentQueue<ChunkRecord> _recentChunks = new();

    public Func<ScannerEvent, Task>? OnScan { get; set; }
    public bool IsMockConnected { get; private set; }
    public bool IsKeyboardWedge { get; private set; }

    // ==================== Configure ====================

    public async Task<object> ConfigureAsync(CommConfig config)
    {
        lock (_lock)
        {
            _currentConfig = new CommConfig
            {
                ComPort = config.ComPort,
                Baudrate = config.Baudrate,
                DataBits = config.DataBits,
                Parity = config.Parity?.ToLower() ?? "none",
                StopBits = config.StopBits,
                TimeoutMs = config.TimeoutMs,
                Dtr = config.Dtr,
                Rts = config.Rts,
                Enabled = config.Enabled
            };

            if (!_currentConfig.Enabled)
            {
                Disconnect();
                IsMockConnected = false;
                IsKeyboardWedge = false;
                return Task.FromResult((object)new { connected = false, enabled = false });
            }

            // Keyboard wedge mode: skip serial, mark ready for USB HID input
            if (string.Equals(_currentConfig.ComPort, "KEYBOARD_WEDGE", StringComparison.OrdinalIgnoreCase))
            {
                Disconnect();
                IsMockConnected = false;
                IsKeyboardWedge = true;
                ResetDebugState();
                Console.WriteLine("[scanner] keyboard wedge mode enabled");
                return Task.FromResult((object)new { connected = true, enabled = true });
            }

            // Mock mode: skip serial, mark connected
            if (string.Equals(_currentConfig.ComPort, "MOCK_SCANNER", StringComparison.OrdinalIgnoreCase))
            {
                if (!_flags.EnableMockMode)
                    throw new InvalidOperationException("Mock mode is not enabled. Cannot use MOCK_SCANNER.");

                Disconnect();
                IsMockConnected = true;
                ResetDebugState();
                Console.WriteLine("[scanner] mock scanner enabled");
                return Task.FromResult((object)new { connected = true, enabled = true });
            }

            IsMockConnected = false;
            IsKeyboardWedge = false;
            ResetDebugState();
        }

        await ConnectAsync();
        return new { connected = IsConnected, enabled = true };
    }

    private async Task ConnectAsync()
    {
        if (_currentConfig == null) return;

        await DisconnectAsync();
        var parity = _currentConfig.Parity?.ToLower() switch
        {
            "even" => Parity.Even, "odd" => Parity.Odd,
            "mark" => Parity.Mark, "space" => Parity.Space,
            _ => Parity.None
        };
        var stopBits = _currentConfig.StopBits switch
        {
            1 => StopBits.One, 1.5 => StopBits.OnePointFive, 2 => StopBits.Two, _ => StopBits.One
        };

        _port = new SerialPort(_currentConfig.ComPort, _currentConfig.Baudrate, parity, _currentConfig.DataBits, stopBits)
        {
            ReadTimeout = _currentConfig.TimeoutMs,
            WriteTimeout = _currentConfig.TimeoutMs
        };
        _port.Open();

        ApplyLineSignals();
        _readCts = new CancellationTokenSource();
        _ = Task.Run(() => ReadLoop(_readCts.Token));

        Console.WriteLine($"[scanner] connected {_currentConfig.ComPort}");
    }

    private async Task ReadLoop(CancellationToken ct)
    {
        var flushTimer = (Timer?)null;
        try
        {
            while (!ct.IsCancellationRequested && _port?.IsOpen == true)
            {
                try
                {
                    if (_port.BytesToRead > 0)
                    {
                        var buf = new byte[Math.Min(_port.BytesToRead, 4096)];
                        var read = _port.Read(buf, 0, buf.Length);
                        if (read > 0)
                            HandleIncomingChunk(buf.Take(read).ToArray(), ref flushTimer);
                    }
                }
                catch (TimeoutException) { /* no data, continue */ }
                catch (ObjectDisposedException) { break; }
                catch (IOException) { break; }
                catch (InvalidOperationException) { break; }
                await Task.Delay(10, ct);
            }
        }
        catch (OperationCanceledException) { }
        finally
        {
            flushTimer?.Dispose();
        }
    }

    private void HandleIncomingChunk(byte[] chunk, ref Timer? flushTimer)
    {
        var textChunk = System.Text.Encoding.UTF8.GetString(chunk);

        lock (_lock)
        {
            _bytesReceived += chunk.Length;
            _chunksReceived++;
            _lastChunkAt = DateTime.UtcNow.ToString("o");

            // Debug chunk recording
            _recentChunks.Enqueue(new ChunkRecord
            {
                At = _lastChunkAt,
                Size = chunk.Length,
                Hex = BitConverter.ToString(chunk).Replace("-", " "),
                TextPreview = textChunk.Replace("\r", "<CR>").Replace("\n", "<LF>")
            });
            while (_recentChunks.Count > DebugChunkLimit)
                _recentChunks.TryDequeue(out _);

            _buffer += textChunk;
            _bufferPreview = _buffer.Replace("\r", "<CR>").Replace("\n", "<LF>");
            if (_bufferPreview.Length > 200)
                _bufferPreview = _bufferPreview[^200..];

            // Parse scan lines
            int boundaryIdx;
            while ((boundaryIdx = FindBoundaryIndex(_buffer)) != -1)
            {
                var payload = _buffer[..boundaryIdx];
                _buffer = _buffer[(boundaryIdx + 1)..];
                PublishScan(payload);
            }
        }

        // Reset flush timer
        flushTimer?.Dispose();
        flushTimer = new Timer(_ =>
        {
            string payload;
            lock (_lock)
            {
                payload = _buffer;
                _buffer = "";
                _bufferPreview = "";
            }
            PublishScan(payload);
        }, null, ScanIdleFlushMs, Timeout.Infinite);
    }

    private static int FindBoundaryIndex(string text)
    {
        if (string.IsNullOrEmpty(text)) return -1;
        var cr = text.IndexOf('\r');
        var lf = text.IndexOf('\n');
        if (cr == -1) return lf;
        if (lf == -1) return cr;
        return Math.Min(cr, lf);
    }

    private void PublishScan(string rawValue)
    {
        var rawText = NormalizeScanText(rawValue);
        if (string.IsNullOrEmpty(rawText)) return;

        lock (_lock)
        {
            _latestScan = new ScannerEvent
            {
                RawText = rawText,
                ScannedAt = DateTime.UtcNow.ToString("o")
            };
            _lastPublishedAt = _latestScan.ScannedAt;
        }

        Console.WriteLine($"[scanner] scan received {rawText}");
        var scan = _latestScan;
        _ = Task.Run(async () =>
        {
            if (OnScan != null) await OnScan(scan);
        });
    }

    private static string NormalizeScanText(string raw)
        => System.Text.RegularExpressions.Regex.Replace(raw ?? "", @"[ -]", "").Trim();

    // ==================== Disconnect ====================

    private void Disconnect()
    {
        _readCts?.Cancel();
        _port?.Close();
        _port?.Dispose();
        _port = null;
        _buffer = "";
        _bufferPreview = "";
    }

    private async Task DisconnectAsync()
    {
        _readCts?.Cancel();
        if (_port != null)
        {
            try { if (_port.IsOpen) _port.Close(); }
            catch { /* ignore */ }
            _port.Dispose();
            _port = null;
        }
        _buffer = "";
        _bufferPreview = "";
        await Task.CompletedTask;
    }

    public bool IsConnected => _port?.IsOpen == true || IsMockConnected || IsKeyboardWedge;

    /// <summary>Accept a scan event from keyboard wedge or mock input.</summary>
    public ScannerEvent AcceptScan(string rawText)
    {
        var scanEvent = new ScannerEvent
        {
            RawText = rawText,
            ScannedAt = DateTime.UtcNow.ToString("o")
        };
        lock (_lock)
        {
            _latestScan = scanEvent;
            _lastPublishedAt = scanEvent.ScannedAt;
            _bytesReceived += rawText.Length;
            _chunksReceived++;
            _lastChunkAt = scanEvent.ScannedAt;
        }
        Console.WriteLine($"[scanner] scan published: {rawText}");

        // Fire the OnScan callback if set (same path as real scan)
        var scan = scanEvent;
        _ = Task.Run(async () =>
        {
            if (OnScan != null) await OnScan(scan);
        });

        return scanEvent;
    }

    private void ResetDebugState()
    {
        _bytesReceived = 0;
        _chunksReceived = 0;
        _lastChunkAt = null;
        _lastPublishedAt = null;
        _modemSignals = null;
        _bufferPreview = "";
        while (_recentChunks.TryDequeue(out _)) { }
    }

    // ==================== Scan management ====================

    public ScannerEvent? GetLatestScan() { lock (_lock) return _latestScan; }

    public void SyncLatestScan(ScannerEvent? scanEvent)
    {
        if (scanEvent?.RawText == null) return;
        lock (_lock)
        {
            if (scanEvent.Id != null && _consumedScanId != null && scanEvent.Id != _consumedScanId)
                _consumedScanId = null;
            _latestScan = new ScannerEvent
            {
                Id = scanEvent.Id ?? "",
                RawText = scanEvent.RawText,
                ScannedAt = scanEvent.ScannedAt ?? DateTime.UtcNow.ToString("o")
            };
            _lastPublishedAt = _latestScan.ScannedAt;
        }
    }

    public bool IsConsumedScan(ScannerEvent? scanEvent)
        => scanEvent?.Id != null && _consumedScanId != null && scanEvent.Id == _consumedScanId;

    public ScannerEvent? GetLatestVisibleScan(ScannerEvent? fallbackScan = null)
    {
        lock (_lock)
        {
            if (_latestScan != null) return _latestScan;
            if (IsConsumedScan(fallbackScan)) return null;
            return fallbackScan;
        }
    }

    public void ClearLatestScan(object? match = null)
    {
        lock (_lock)
        {
            if (_latestScan == null) return;
            if (match == null) { _latestScan = null; return; }

            // Try to match by properties (reflection-light approach)
            var matchDict = match.GetType().GetProperties()
                .ToDictionary(p => p.Name, p => p.GetValue(match)?.ToString());
            var scanEventId = matchDict.GetValueOrDefault("ScannerEventId");
            var qrCode = matchDict.GetValueOrDefault("QrCode");

            var idMatched = scanEventId != null && _latestScan.Id == scanEventId;
            var qrMatched = _latestScan.Id == null && qrCode != null && _latestScan.RawText == qrCode;
            if (idMatched || qrMatched) _latestScan = null;
        }
    }

    public void MarkScanConsumed(object? match = null)
    {
        lock (_lock)
        {
            var matchDict = match?.GetType().GetProperties()
                .ToDictionary(p => p.Name, p => p.GetValue(match)?.ToString());
            if (matchDict?.TryGetValue("ScannerEventId", out var scanEventId) == true && scanEventId != null)
                _consumedScanId = scanEventId;
            ClearLatestScan(match);
        }
    }

    public void ConsumeCurrentScan(object? match = null)
    {
        lock (_lock)
        {
            var matchDict = match?.GetType().GetProperties()
                .ToDictionary(p => p.Name, p => p.GetValue(match)?.ToString());
            if (matchDict?.TryGetValue("ScannerEventId", out var scanEventId) == true && scanEventId != null)
                _consumedScanId = scanEventId;
            else if (_latestScan?.Id != null)
                _consumedScanId = _latestScan.Id;

            _latestScan = null;
            _buffer = "";
            _bufferPreview = "";
        }
    }

    // ==================== Debug ====================

    public object GetDebugState()
    {
        var chunks = _recentChunks.ToArray();
        return new
        {
            connected = IsConnected,
            latestScan = _latestScan,
            currentConfig = _currentConfig,
            debug = new
            {
                bytesReceived = _bytesReceived,
                chunksReceived = _chunksReceived,
                lastChunkAt = _lastChunkAt,
                lastPublishedAt = _lastPublishedAt,
                modemSignals = _modemSignals,
                bufferPreview = _bufferPreview,
                recentChunks = chunks
            }
        };
    }

    public async Task<object> UpdateLineSignalsAsync(object options)
    {
        var opt = options.GetType().GetProperties().ToDictionary(p => p.Name, p => p.GetValue(options));
        var dtr = opt.TryGetValue("Dtr", out var v) && v is bool bDtr ? bDtr : true;
        var rts = opt.TryGetValue("Rts", out var rv) && rv is bool bRts ? bRts : true;
        var reconnect = !opt.TryGetValue("Reconnect", out var rcv) || rcv is not bool bRec || bRec;

        if (_currentConfig == null)
            throw new InvalidOperationException("Scanner config not initialized");

        _currentConfig.Dtr = dtr;
        _currentConfig.Rts = rts;

        if (reconnect)
            await ConnectAsync();
        else if (_port?.IsOpen == true)
            ApplyLineSignals();

        return GetDebugState();
    }

    private void ApplyLineSignals()
    {
        if (_port?.IsOpen != true || _currentConfig == null) return;

        _port.DtrEnable = _currentConfig.Dtr;
        _port.RtsEnable = _currentConfig.Rts;

        _modemSignals = new
        {
            CtsHolding = _port.CtsHolding,
            DsrHolding = _port.DsrHolding,
            CDHolding = _port.CDHolding
        };
    }

    private record ChunkRecord
    {
        public string? At { get; set; }
        public int Size { get; set; }
        public string Hex { get; set; } = "";
        public string TextPreview { get; set; } = "";
    }
}
